### PR TITLE
[MNG-7774] Maven config and command line interpolation

### PR DESCRIPTION
**Pull request by cstamas**
Reuse as much as possible from master, but keep existing stuff like multiModuleProjectDirectory alone.

Changes:
* interpolate user properties and arguments
* introduce session.topDirectory and session.rootDirectory expressions (for interpolation only)
* Maven fails to start if any of the new properties are undefined but their use is attempted
* leave everything else untouched

---

https://issues.apache.org/jira/browse/MNG-7774

### CODE REVIEW

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 1]
```diff
@@ -140,9 +146,14 @@ public class MavenCli {
...
     private static final String EXT_CLASS_PATH = "maven.ext.class.path";
 
-    private static final String EXTENSIONS_FILENAME = ".mvn/extensions.xml";
+    private static final String DOT_MVN = ".mvn";
 
-    private static final String MVN_MAVEN_CONFIG = ".mvn/maven.config";
+    private static final String UNABLE_TO_FIND_ROOT_PROJECT_MESSAGE = "Unable to find the root directory. Create a "
+            + DOT_MVN + " directory in the project root directory to identify it.";
+
+    private static final String EXTENSIONS_FILENAME = DOT_MVN + "/extensions.xml";
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:154]: That's not a filename anymore, but a path.

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 2]
```diff
@@ -140,9 +146,14 @@ public class MavenCli {
...
+
+    private static final String MVN_MAVEN_CONFIG = DOT_MVN + "/maven.config";
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:156]: Suffers from the same more or less.

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 3]
```diff
@@ -526,8 +578,39 @@ private void commands(CliRequest cliRequest) {
 
     // Needed to make this method package visible to make writing a unit test possible
     // Maybe it's better to move some of those methods to separate class (SoC).
-    void properties(CliRequest cliRequest) {
-        populateProperties(cliRequest.commandLine, cliRequest.systemProperties, cliRequest.userProperties);
+    void properties(CliRequest cliRequest) throws ExitException {
+        try {
+            populateProperties(cliRequest, cliRequest.systemProperties, cliRequest.userProperties);
+
+            StringSearchInterpolator interpolator =
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: Any reason not to use the `StringRegexInterpolator`?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 1 above)
Comment (cstamas) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:154]: Before and after this PR the contents of `EXTENSIONS_FILENAME` is unchanged, so I take this as nitpicking :smile:

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 3 above)
Comment (cstamas) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: The interpolator thing is pretty much coming from master. What is the difference?

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 1 above)
Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:154]: Shit, you are right. It just popped my eye.

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java (hunk 3 above)
Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: I remember that one has superseded the other, I don't exactly remember.

Comment (cstamas) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: As I see in maven codebase use of StringSearch vs Regexp interpolator is 4:5

Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: I wonder they there are two what the difference is.

Comment (cstamas) [reply maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:585]: Well, I remember the interface https://github.com/codehaus-plexus/plexus-interpolation/blob/master/src/main/java/org/codehaus/plexus/interpolation/Interpolator.java#L24 :smile:

File: maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java [hunk 4]
```diff
@@ -309,6 +320,47 @@ void initialize(CliRequest cliRequest) throws ExitException {
...
+                break;
+            } else {
+                // Check if this is the -f/--file option
+                isAltFile = arg.equals(String.valueOf(CLIManager.ALTERNATE_POM_FILE)) || arg.equals("file");
+            }
+        }
+        try {
+            topDirectory = topDirectory.toAbsolutePath().toRealPath();
+        } catch (IOException e) {
+            System.err.println("Error computing real path from " + topDirectory);
```
Comment (CrazyHZM) [inline maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:355]: Would it be easier to troubleshoot the problem if the exception information was printed out here?

**Comment by cstamas** [comment]
> Does this actually supersede #1062?

No, that one is for master and builds on top of already present new mvn4 features not present in 3.9.x. This PR is for 3.9.x and "mildly follows" the one you refer to, with reduced functionalities. For example, that one uses `RootLocator` (already on master) while we have no such thing. Master also uses two strategies to identify "root", but it requires consumer POM feature, again not available here, etc. This PR provides "minimum viable change" to provide in 3.9.x:
* config and command line interpolation
* two new properties (but for interpolation ONLY), unlike on master, where these two bubble up to session and project
* the UT is intentionally 1:1 (so it IS copy pasted from referenced PR)

**Comment by cstamas** [comment]
Anyone else?

**Comment by cstamas** [comment]
> @cstamas I think the description of adding this command in `mvn -help` is still missing.

Which command? No command was added or modified in Maven CLI

**Comment by CrazyHZM** [comment]
> > @cstamas I think the description of adding this command in `mvn -help` is still missing.
> 
> Which command? No command was added or modified in Maven CLI

My mistake was that I didn't see it clearly. There was nothing wrong with 3.9.x. There are too many maven versions for my local development.  😢   @cstamas

**Review by CrazyHZM — APPROVED** [review]
LGTM

File: maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java [hunk 5]
```diff
@@ -364,6 +369,38 @@ public void testVersionStringWithoutAnsi() throws Exception {
...
+    @Test
+    public void testPropertiesInterpolation() throws Exception {
+        // Arrange
+        CliRequest request = new CliRequest(
+                new String[] {
+                    "-Dfoo=bar",
+                    "-DvalFound=s${foo}i",
+                    "-DvalNotFound=s${foz}i",
+                    "-DvalRootDirectory=${session.rootDirectory}/.mvn/foo",
+                    "-DvalTopDirectory=${session.topDirectory}/pom.xml",
```
Comment (michael-o) [inline maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java:381]: I have a problem with these two expressions: They haven't been back ported to 3.9.x, no? So they are fake here and not work throughout the entire build?

Comment (cstamas) [reply maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java:381]: Expressions HAVE BEEN (in this PR) backported, what you *really mean* is that 3.9.x does not expose them on MavenSession object, that is true. Still, the point of this PR is to make builds between 3.9.x and 4 permeable, "work with both". This PR as explained above, *does not back port everything*, just the ability to interpolate and these two "special" expressions used in interpolation.

Comment (michael-o) [reply maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java:381]: This I understand now. Hope they won't be misunderstood by users.

**Review by michael-o — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #9211

**Pull request merged**